californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java

```diff
@@ -397,35 +477,41 @@ public void sendResponse(final Exchange exchange, final Response response) {
 	@Override
 	public void receiveResponse(final Exchange exchange, final Response response) {
 
-		if (exchange.getRequest().isCanceled()) {
-			// do not continue fetching blocks if canceled
-			// reject (in particular for Block+Observe)
-			if (response.getType()!=Type.ACK) {
-				LOGGER.finer("rejecting blockwise transfer for canceled Exchange");
-				EmptyMessage rst = EmptyMessage.newRST(response);
-				sendEmptyMessage(exchange, rst);
-				// Matcher sets exchange as complete when RST is sent
-			}
+		if (isTransparentBlockwiseHandlingEnabled()) {
 
-		} else if (!response.hasBlockOption()) {
+			if (exchange.getRequest().isCanceled()) {
```

Comment:
IMHO we should stop an ongoing outbound block1 or inbound block2 transfer if the original request got canceled by the application layer and clear the blockwise status accordingly.
I agree with @boaks that we do not need to send an RST here. If we receive a response (non-empty) in this method then it contains a block of the body or an error code and we can simply clear the block1 and/or block2 status and forget about the request (the application layer is not interested in the response anymore).
However, we will probably need to deal with empty responses we receive for block requests we have sent out subsequent to the initial request. We will need to e.g. mark the original request as rejected.